kotlinx-coroutines-core/jvm/test/LimitedParallelismStressTest.kt

```diff
@@ -0,0 +1,77 @@
+/*
+ * Copyright 2016-2021 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license.
+ */
+
+package kotlinx.coroutines
+
+import org.junit.*
+import org.junit.Test
+import org.junit.runner.*
+import org.junit.runners.*
+import java.util.concurrent.atomic.*
+import kotlin.test.*
+
+@RunWith(Parameterized::class)
+class LimitedParallelismStressTest(private val targetParallelism: Int) : TestBase() {
+
+    companion object {
+        @Parameterized.Parameters(name = "{0}")
+        @JvmStatic
+        fun params(): Collection<Array<Any>> = listOf(1, 2, 3, 4).map { arrayOf(it) }
+    }
+
+    @get:Rule
+    val executor = ExecutorRule(targetParallelism * 2)
+    private val iterations = 100_000 * stressTestMultiplier
+
+    private val parallelism = AtomicInteger(0)
+
+    private fun checkParallelism() {
+        val value = parallelism.incrementAndGet()
+        assertTrue { value <= targetParallelism }
```

Comment:
On my Linux machine, `100_000` iterations weren't enough to reach even the target parallelism of 4. Maybe this means that some artificial delay should be inserted here, but it's your call.

Reply:
I've added `Thread.yield`, do you mind rechecking this?

In general, my approach to stress tests is the following:

* Write some generic tests that just stress out the entity
* I usually have a clear mental model of the entity being tested. I do write tests for the target's potential weak points and easy-to-make bugs according to my model. Then I **do** make such bugs and ensure that at least one of the tests fails

So for this test I've ensured that at least one test fails if I let the bug slip to `LImitedDispatcher`, though not sure about each parallelism level